src/app/shared/services/command.service.ts

```diff
@@ -0,0 +1,295 @@
+﻿import { DestroyRef, inject, Injectable } from '@angular/core';
+import { ChatCommand, ChoiceCommand, Commands, CounterCommand, SoundCommand } from './command.interface';
+import { LogService } from './logs.service';
+import { UserPermissions } from './chat.interface';
+import { VStreamService } from './vstream.service';
+import { filter, first, map, tap } from 'rxjs';
+import { ChatService } from './chat.service';
+import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class CommandService {
+  private readonly vstreamService = inject(VStreamService);
+  private readonly chatService = inject(ChatService);
+  private readonly logService = inject(LogService);
+  private readonly destroyRef = inject(DestroyRef);
+
+  private readonly commands$ = this.vstreamService.commands$;
+
+  readonly cooldowns = new Map<string, {
+    duration: number,
+    onCooldown: boolean,
+    timeout?: NodeJS.Timeout,
+  }>();
+
+  readonly autoredeems = new Map<string, {
+    interval: number,
+    timer: NodeJS.Timer,
+  }>();
+
+  constructor() {
+    this.commands$
+      .pipe(
+        takeUntilDestroyed(),
+        tap(commands => {
+          const commandIDs = commands.map(c => c.id);
+          const chatCommands = commands.filter((c): c is ChatCommand => c.type === 'chat');
+          this.clearOldAutoIntervals(commandIDs);
+          this.resetAutoIntervals(chatCommands);
+          this.updateCooldowns(commands);
+        }),
+      ).subscribe();
+  }
+
+  /**
+   * If a user has updated the cooldowns of an existing command, we want to remove any active cooldowns.
+   * @param commands All the VStream chat commands
+   * @private
+   */
+  private updateCooldowns(commands: Commands[]) {
+    for (const command of commands) {
+      const cooldown = this.cooldowns.get(command.id);
+
+      if (!cooldown) {
+        continue;
+      }
+
+      const { duration, timeout } = cooldown;
+      const newDuration = command.cooldown * 1000;
+
+      if (duration === newDuration) {
+        continue;
+      }
+
+      clearTimeout(timeout);
+      this.cooldowns.delete(command.id);
+    }
+  }
+
+  private resetAutoIntervals(commands: ChatCommand[]) {
+    /**
+     * I want to be safe and have the lowest interval be 1 minute.
+     */
+    const LOWEST_INTERVAL = 60 * 1000;
+
+    for (const command of commands) {
+      const interval = command.autoRedeemInterval * LOWEST_INTERVAL;
+      const isIntervalValid = interval >= LOWEST_INTERVAL;
+      const autoRedeem = this.autoredeems.get(command.id);
+
+      if (command.autoRedeem && !autoRedeem && isIntervalValid) {
+        this.autoredeems.set(
+          command.id,
+          {
+            interval,
+            timer: setInterval(() => this.handleAutoRedeem(command.id), interval),
```

Comment:
If you felt like being more rxjs'y, you could always do `timer: timer(interval).subscribe(() => this.handle...` then when you want to end the timer early just do `redeem.get(id).timer.unsubscribe()`